core_api/src/routes/chat.py

```diff
@@ -162,3 +168,125 @@ def rag_chat(chat_request: ChatRequest, _user_uuid: Annotated[UUID, Depends(get_
         for langchain_document in result.get("input_documents", [])
     ]
     return ChatResponse(output_text=result["output_text"], source_documents=source_documents)
+
+
+@chat_app.websocket("/rag-stream")
+async def websocket_endpoint(websocket: WebSocket):
+    await websocket.accept()
+
+    prompt_search_query = ChatPromptTemplate.from_messages(
+        [
+            MessagesPlaceholder(variable_name="chat_history"),
+            ("user", "{input}"),
+            (
+                "user",
+                "Given the above conversation, generate a search query to look up to get information relevant to the conversation",
+            ),
+        ]
+    )
+
+    retriever_chain = create_history_aware_retriever(llm, vector_store.as_retriever(), prompt_search_query)
+
+    prompt_get_answer = ChatPromptTemplate.from_messages(
+        [
+            ("system", "Answer the user's questions based on the below context:\\n\\n{context}"),
+            MessagesPlaceholder(variable_name="chat_history"),
+            ("user", "{input}"),
+        ]
+    )
+
+    document_chain = create_stuff_documents_chain(llm, prompt_get_answer)
+    retrieval_chain = create_retrieval_chain(retriever_chain, document_chain)
+
+    while True:
+        chat_request = ChatRequest.parse_raw(await websocket.receive_text())
+        chat_history = [
+            HumanMessage(content=x.text) if x.role == "user" else AIMessage(content=x.text)
+            for x in chat_request.message_history[:-1]
+        ]
+        chat = {"chat_history": chat_history, "input": chat_request.message_history[-1].text}
+
+        async for event in retrieval_chain.astream_events(chat, version="v1"):
+            kind = event["event"]
+            if kind == "on_chat_model_stream":
+                await websocket.send_text(event["data"]["chunk"].content)
```

Comment:
Is this where we'll be wrapping the chunk on JSON?

Reply:
yep, i guess so

Reply:
Happy for me to jump in and take a run at it this afternoon?

Reply:
Can we wait until after planning? I want to break this one down a bit and farm out the pieces rather than tackle this in one go

Reply:
Sounds sensible.